File: Fritz TR064/keychain.m
//
//  keychain.m
//
//
//  Created by Troutslayer33 on 7/13/15.
//  Copyright (c) 2015 Troutslayer33. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "keychain.h"
#import "UIKit/UIKit.h"

@implementation keychain : NSObject

+ (void) storeData: (NSString * )key data:(NSData *)data {
    NSMutableDictionary * dict = [[NSMutableDictionary alloc] init];
    [dict setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    NSData *encodedKey = [key dataUsingEncoding:NSUTF8StringEncoding];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrGeneric];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrAccount];
    [dict setObject:@"FritzVPN" forKey:(__bridge id)kSecAttrService];
    [dict setObject:(__bridge id)kSecAttrAccessibleAlwaysThisDeviceOnly forKey:(__bridge id)kSecAttrAccessible];
    [dict setObject:data forKey:(__bridge id)kSecValueData];
    
    OSStatus status = SecItemAdd((__bridge CFDictionaryRef)dict, NULL);
    if(errSecSuccess != status) {
        NSLog(@"Unable add item with key =%@ error:%d",key,(int)status);
    } else {

    }
}

+ (NSData *) getData: (NSString *)key {
    NSMutableDictionary * dict = [[NSMutableDictionary alloc] init];
    [dict setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    NSData *encodedKey = [key dataUsingEncoding:NSUTF8StringEncoding];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrGeneric];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrAccount];
    [dict setObject:@"FritzVPN" forKey:(__bridge id)kSecAttrService];
    [dict setObject:(__bridge id)kSecAttrAccessibleAlwaysThisDeviceOnly forKey:(__bridge id)kSecAttrAccessible];
    [dict setObject:(__bridge id)kSecMatchLimitOne forKey:(__bridge id)kSecMatchLimit];
    [dict setObject:(id)kCFBooleanTrue forKey:(__bridge id)kSecReturnPersistentRef];
    
    CFTypeRef result = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)dict,&result);
    
    if( status != errSecSuccess) {
        NSLog(@"Unable to fetch item for key %@ with error:%d",key,(int)status);
        return nil;
    }
    
    NSData *resultData = (__bridge NSData *)result;
    return resultData;
}

+ (void)deleteKeychainValue:(NSString *)key {
    
    NSMutableDictionary * dict = [[NSMutableDictionary alloc] init];
    [dict setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    NSData *encodedKey = [key dataUsingEncoding:NSUTF8StringEncoding];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrGeneric];
    [dict setObject:encodedKey forKey:(__bridge id)kSecAttrAccount];
    [dict setObject:@"VPN" forKey:(__bridge id)kSecAttrService];
    [dict setObject:(__bridge id)kSecAttrAccessibleAlwaysThisDeviceOnly forKey:(__bridge id)kSecAttrAccessible];
    
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)dict);
    if( status != errSecSuccess) {
        NSLog(@"Unable to fetch item for key %@ with error:%d",key,(int)status);

    } else {

    }

}


@end

File: Fritz TR064/Keychain.swift
//
//  Keychain.swift
//  Fritz TR064
//
//  Created by Daniel Müllenborn on 25/11/15.
//  Copyright © 2015 Daniel Müllenborn. All rights reserved.
//

import Foundation

class Keychain {
  
  private static let SecMatchLimit = kSecMatchLimit as String,
    SecReturnData = kSecReturnData as String,
    SecReturnPersistentRef: String! = kSecReturnPersistentRef as String,
    SecValueData = kSecValueData as String,
    SecAttrAccessible = kSecAttrAccessible as String,
    SecClass = kSecClass as String,
    SecAttrService = kSecAttrService as String,
    SecAttrGeneric = kSecAttrGeneric as String,
    SecAttrAccount = kSecAttrAccount as String

  static func stringForKey(keyName: String) -> String? {
    return self.dataForKey(keyName).flatMap { NSString(data: $0, encoding: NSUTF8StringEncoding) as String? }
  }
  
  static func dataForKey(keyName: String) -> NSData? {
    var keychainQuery = self.setupKeychainQueryForKey(keyName)
    var result: AnyObject?
    
    keychainQuery[SecMatchLimit] = kSecMatchLimitOne
    keychainQuery[SecReturnData] = kCFBooleanTrue
    
    let status = withUnsafeMutablePointer(&result) {
      SecItemCopyMatching(keychainQuery, $0)
    }
    
    return status == noErr ? result as? NSData : nil
  }
  
  static func persistentRefForKey(keyName: String) -> NSData? {
    var keychainQuery = self.setupKeychainQueryForKey(keyName)
    var result: AnyObject?
    
    keychainQuery[SecMatchLimit] = kSecMatchLimitOne
    keychainQuery[SecReturnPersistentRef] = kCFBooleanTrue
    
    let status = withUnsafeMutablePointer(&result) {
      SecItemCopyMatching(keychainQuery, $0)
    }
    
    return status == noErr ? result as? NSData : nil
  }
  
  static func setString(value: String, forKey keyName: String) -> Bool {
    if let data = value.dataUsingEncoding(NSUTF8StringEncoding) {
      return self.setData(data, forKey: keyName)
    } else {
      return false
    }
  }
  
  static func setData(value: NSData, forKey keyName: String) -> Bool {
    var keychainQuery: [String:AnyObject] = self.setupKeychainQueryForKey(keyName)
    
    keychainQuery[SecValueData] = value
    keychainQuery[SecReturnPersistentRef] = kCFBooleanTrue
    let status: OSStatus = SecItemAdd(keychainQuery, nil)
    
    if status == errSecSuccess {
      return true
    } else if status == errSecDuplicateItem {
      return self.updateData(value, forKey: keyName)
    } else {
      return false
    }
  }
  
  static func removeObjectForKey(keyName: String) -> Bool {
    let keychainQuery: [String:AnyObject] = self.setupKeychainQueryForKey(keyName)
    
    let status: OSStatus =  SecItemDelete(keychainQuery);
    
    if status == errSecSuccess {
      return true
    } else {
      return false
    }
  }
  
  private static func updateData(value: NSData, forKey keyName: String) -> Bool {
    let keychainQuery: [String:AnyObject] = self.setupKeychainQueryForKey(keyName)
    let updateDictionary = [SecValueData:value]

    let status: OSStatus = SecItemUpdate(keychainQuery, updateDictionary)
    
    if status == errSecSuccess {
      return true
    } else {
      return false
    }
  }
  
  private static func setupKeychainQueryForKey(keyName: String) -> [String:AnyObject] {
    var keychainQuery: [String:AnyObject] = [SecClass:kSecClassGenericPassword]
    
    let encodedIdentifier: NSData? = keyName.dataUsingEncoding(NSUTF8StringEncoding)
    
    keychainQuery[SecAttrGeneric] = encodedIdentifier
    keychainQuery[SecAttrAccount] = encodedIdentifier
    
    keychainQuery[SecAttrService] = "FritzVPN"
    
    keychainQuery[SecAttrAccessible] = kSecAttrAccessibleAlwaysThisDeviceOnly
    
    return keychainQuery
  }
}


